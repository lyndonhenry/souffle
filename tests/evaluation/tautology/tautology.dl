
#define TRUE_NUMBER 1
#define TRUE_SYMBOL "1"

.type Symbol = symbol
.type Number = number

.comp Nullary {
    .decl truth()
    .decl falsehood()
    truth().
}

.comp Unary<Type> {
    .decl truth(x:Type)
    .decl falsehood(x:Type)
}

.comp UnaryNumber : Unary<Number> {
    truth(TRUE_NUMBER).
}

.comp UnarySymbol : Unary<Symbol> {
    truth(TRUE_SYMBOL).
}

.comp Binary<Type> {
    .decl truth(x:Type, y:Type)
    .decl falsehood(x:Type, y:Type)
}

.comp BinaryNumber : Binary<Number> {
    truth(TRUE_NUMBER, TRUE_NUMBER).
}

.comp BinarySymbol : Binary<Symbol> {
    truth(TRUE_SYMBOL, TRUE_SYMBOL).
}

.init nullary = Nullary
.init unary_number = UnaryNumber
.init unary_symbol = UnarySymbol
.init binary_number = BinaryNumber
.init binary_symbol = BinarySymbol

.decl tautology()
.printsize tautology
.output tautology

tautology() :-
    nullary.truth(),
   !nullary.falsehood(),
    unary_number.truth(_),
   !unary_number.falsehood(_),
    unary_symbol.truth(_),
   !unary_symbol.falsehood(_),
    binary_number.truth(_, _),
   !binary_number.falsehood(_, _),
    binary_symbol.truth(_, _),
   !binary_symbol.falsehood(_, _).

.decl contradiction()
.printsize contradiction

contradiction() :-
   !nullary.truth(),
  !!nullary.falsehood(),
   !unary_number.truth(_),
  !!unary_number.falsehood(_),
   !unary_symbol.truth(_),
  !!unary_symbol.falsehood(_),
   !binary_number.truth(_, _),
  !!binary_number.falsehood(_, _),
   !binary_symbol.truth(_, _),
  !!binary_symbol.falsehood(_, _).
