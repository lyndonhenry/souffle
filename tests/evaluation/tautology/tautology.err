Warning: No rules/facts defined for relation nullary.falsehood in file tautology.dl at line 10
    .decl falsehood()
----------^-----------
Warning: No rules/facts defined for relation unary_number.falsehood in file tautology.dl at line 16
    .decl falsehood(x:Type)
----------^-----------------
Warning: No rules/facts defined for relation unary_symbol.falsehood in file tautology.dl at line 16
    .decl falsehood(x:Type)
----------^-----------------
Warning: No rules/facts defined for relation binary_number.falsehood in file tautology.dl at line 29
    .decl falsehood(x:Type, y:Type)
----------^-------------------------
Warning: No rules/facts defined for relation binary_symbol.falsehood in file tautology.dl at line 29
    .decl falsehood(x:Type, y:Type)
----------^-------------------------