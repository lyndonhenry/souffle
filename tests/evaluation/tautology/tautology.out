contradiction	0
tautology	1
